Add docs and tests for infeasibility certificates (#1660)
jump-dev/MathOptInterface.jl: docs/src/background/infeasibility_certificates.md

here is a brief overview.

For a minimization problem in geometric conic form, the primal is:
```math
\begin{align}
& \min_{x \in \mathbb{R}^n} & a_0^\top x + b_0
\\
& \;\;\text{s.t.} & A_i x + b_i & \in \mathcal{C}_i & i = 1 \ldots m
\end{align}
```
and the dual is a maximization problem in standard conic form:
```math
\begin{align}
& \max_{y_1, \ldots, y_m} & -\sum_{i=1}^m b_i^\top y_i + b_0
\\
& \;\;\text{s.t.} & a_0 - \sum_{i=1}^m A_i^\top y_i & = 0
\\
& & y_i & \in \mathcal{C}_i^* & i = 1 \ldots m
\end{align}
```
where each ``\mathcal{C}_i`` is a closed convex cone and ``\mathcal{C}_i^*`` is
its dual cone.

For a maximization problem in geometric conic form, the primal is:
```math
\begin{align}
& \max_{x \in \mathbb{R}^n} & a_0^\top x + b_0
\\
& \;\;\text{s.t.} & A_i x + b_i & \in \mathcal{C}_i & i = 1 \ldots m
\end{align}
```
and the dual is a minimization problem in standard conic form:
```math
\begin{align}
& \min_{y_1, \ldots, y_m} & \sum_{i=1}^m b_i^\top y_i + b_0
\\
& \;\;\text{s.t.} & a_0 + \sum_{i=1}^m A_i^\top y_i & = 0
\\
& & y_i & \in \mathcal{C}_i^* & i = 1 \ldots m
\end{align}
```

## Unbounded models

> For an unbounded model, a certificate of infeasibility is an improving ray of

Reviewer: "or an unbounded model, a certificate of infeasibility" can be confusing as it's not clear whether we're certifying unboundness or infeasiblity. I would add something like that:

For an unbounded model, a certificate of unboundedness is made of two separate certificates:
1) A feasible solution `x0` certifying that the primal is feasible
2) A direction `d` that is both feasible for the homogenized feasible set and is a direction of decrease for the objective (in case of minimization).

Combining these two certificates, by convexity of the feasible set and linearity of the objective function, we can see that `x0 + \eta d` is feasible for any \eta >= 0 and can have an objective value arbitrarily small for increasing values of \eta.

The direction `d` is also a certificate of infeasibility of the dual problem.
In fact, most solvers stop when they have found `d` and do not search for a feasible solution `x0` so they terminate with a complete certificate of dual infeasibility but an incomplete certificate of primal unboundedness.
In that situation, the dual is known to be infeasible but the primal can still be either infeasible or unbounded.
The two cases can be distinguished by setting the `ObjectiveSense` to `FEASIBILITY_SENSE` and calling `optimize!` on the modified model.

Author: Yeah I should just call it primal and dual infeasibility certificates and add a note about solving with no objective.

the primal problem. That is, there exists some vector ``d`` such that for all
``\eta > 0``:
```math
A_i (x + \eta d) + b_i \in \mathcal{C}_i,\ \ i = 1 \ldots m,
```
and (for minimization problems):
```math
a_0^\top (x + \eta d) + b_0 \le a_0^\top x + b_0
```

> for any feasible point ``x``. The latter simplifies to ``a_0^\top d \le 0``. For

Reviewer: ```suggestion
for any feasible point ``x``. The latter simplifies to ``a_0^\top d < 0``. For
```

> maximization problems, the inequality is reversed, so that ``a_0^\top d \ge 0``.

Reviewer: ```suggestion
maximization problems, the inequality is reversed, so that ``a_0^\top d > 0``.
```

If the solver has found a certificate of infeasibility:

 * [`TerminationStatus`](@ref) must be `DUAL_INFEASIBLE`
 * [`PrimalStatus`](@ref) must be `INFEASIBILITY_CERTIFICATE`
 * [`VariablePrimal`](@ref) must be the corresponding value of ``d``

>  * [`ObjectiveValue`](@ref) must be the value ``a_0^\top d``.

Reviewer: ```suggestion
 * [`ObjectiveValue`](@ref) must be the value ``a_0^\top d``. Note that this is the value of the objective function at `d` ignoring the constant `b_0`.
```

## Infeasible models

For an infeasible model, a certificate of infeasibility is an improving ray of
the dual problem. However, because infeasibility is independent of the objective

> function, we first homogenize the primal problem by removing the objective.

Reviewer: ```suggestion
function, we first homogenize the primal problem by removing the primal objective.
```

Reviewer: It's unclear whether it's the objective of the primal or dual

Therefore, a dual improving ray is some vector ``d`` such that for all
``\eta > 0``:
```math
\begin{align}
\sum_{i=1}^m A_i^\top (y_i + \eta d_i) & = 0 \\
(y_i + \eta d_i) & \in \mathcal{C}_i^* & i = 1 \ldots m,
\end{align}
```
and:
```math

> \sum_{i=1}^m b_i^\top (y_i + \eta d_i) \le \sum_{i=1}^m b_i^\top y_i

Reviewer: ```suggestion
\sum_{i=1}^m b_i^\top (y_i + \eta d_i) < \sum_{i=1}^m b_i^\top y_i
```

```
for any feasible dual solution ``y``. The latter simplifies to

> ``\sum_{i=1}^m b_i^\top d_i \le 0``. Note that conic duality is defined such

Reviewer: ```suggestion
``\sum_{i=1}^m b_i^\top d_i < 0``. Note that conic duality is defined such
```

> that this inequality is independent the problem being minmization or

Reviewer: ```suggestion
that this inequality is independent of whether the objective sense being minimization or
```

If a model is unbounded, two things are true:
 1. there exists a feasible primal solution
 2. the dual is infeasible.

A feasible primal solution---if one exists---can be obtained by setting
[`ObjectiveSense`](@ref) to `FEASIBILITY_SENSE` and then optimizing. Therefore,
most solvers terminate after they prove the dual is infeasible via a certificate
of dual infeasibility, but _before_ they have found a feasible primal solution.
This is also the reason that MathOptInterface defines the `DUAL_INFEASIBLE`
status instead of `UNBOUNDED`.

A certificate of dual infeasibility is an improving ray of the primal problem.
That is, there exists some vector ``d`` such that for all ``\eta > 0``:
```math
A_i (x + \eta d) + b_i \in \mathcal{C}_i,\ \ i = 1 \ldots m,
```
and (for minimization problems):
```math
a_0^\top (x + \eta d) + b_0 < a_0^\top x + b_0,
```
for any feasible point ``x``. The latter simplifies to ``a_0^\top d < 0``. For
maximization problems, the inequality is reversed, so that ``a_0^\top d > 0``.

If the solver has found a certificate of dual infeasibility:

 * [`TerminationStatus`](@ref) must be `DUAL_INFEASIBLE`
 * [`PrimalStatus`](@ref) must be `INFEASIBILITY_CERTIFICATE`

>  * [`VariablePrimal`](@ref) must be the corresponding value of ``d``

Author: The open question from #433 is what to do with `ConstraintPrimal`. For consistency, it makes the most sense for it to be `A_i d` (i.e., without the constant term `b_i`).

However, I'm not sure if solvers actually support querying this, so it's effort for solver wrappers to implement.

Alternatively, we could just say that querying `ConstraintPrimal` for an infeasibility certificate is undefined behavior.

Author: Note for myself tomorrow: it's actually easy to implement if we update the `get_fallback`, so there's no reason not to explicitly define `ConstraintPrimal` here as `A_i d`.

Reviewer: Yes, `get_fallback` is the way and the constant should not be there, see also https://github.com/jump-dev/MathOptInterface.jl/issues/1310.

Author: Okay done. I still need to add tests for the scalarize and vectorize bridges.

    to the solver.

## Infeasible models

A certificate of primal infeasibility is an improving ray of the dual problem.
However, because infeasibility is independent of the objective function, we
first homogenize the primal problem by removing its objective.

For a minimization problem, a dual improving ray is some vector ``d`` such that
for all ``\eta > 0``:
```math
\begin{align}
\sum_{i=1}^m -A_i^\top (y_i + \eta d_i) & = 0 \\
(y_i + \eta d_i) & \in \mathcal{C}_i^* & i = 1 \ldots m,
\end{align}
```
and:
```math
-\sum_{i=1}^m b_i^\top (y_i + \eta d_i) > -\sum_{i=1}^m b_i^\top y_i,
```
for any feasible dual solution ``y``. The latter simplifies to
``-\sum_{i=1}^m b_i^\top d_i > 0``. For a maximization problem, the inequality
is ``\sum_{i=1}^m b_i^\top d_i < 0``. (Note that these are the same inequality,
modulo a `-` sign.)

If the solver has found a certificate of infeasibility:

 * [`TerminationStatus`](@ref) must be `INFEASIBLE`
 * [`DualStatus`](@ref) must be `INFEASIBILITY_CERTIFICATE`
 * [`ConstraintDual`](@ref) must be the corresponding value of ``d``
 * [`DualObjectiveValue`](@ref) must be the value
   ``-\sum_{i=1}^m b_i^\top d_i`` for minimization problems and

>    ``\sum_{i=1}^m b_i^\top d_i`` for maximization problems.

Author: There's also a question about what to do here. Currently, the dual fallback returns different signs, so that's what I've left it as.

Author: I've convinced myself that this is correct.
